prog/vnet/update_firewall_rules.rb

```diff
@@ -64,38 +67,79 @@ class Prog::Vnet::UpdateFirewallRules < Prog::Base
     type filter hook forward priority filter; policy drop;
     ip protocol tcp counter flow offload @ubi_flowtable
     ip protocol udp counter flow offload @ubi_flowtable
-    ip saddr @private_ipv4_ips ct state established,related,new counter accept
-    ip daddr @private_ipv4_ips ct state established,related counter accept
-    ip6 saddr @private_ipv6_ips ct state established,related,new counter accept
-    ip6 daddr @private_ipv6_ips ct state established,related counter accept
+    ip saddr @private_ipv4_cidrs ct state established,related,new counter accept
+    ip daddr @private_ipv4_cidrs ct state established,related counter accept
+    ip6 saddr @private_ipv6_cidrs ct state established,related,new counter accept
+    ip6 daddr @private_ipv6_cidrs ct state established,related counter accept
     ip6 saddr #{guest_ephemeral} ct state established,related,new counter accept
     ip6 daddr #{guest_ephemeral} ct state established,related counter accept
-    ip saddr @allowed_ipv4_ips ip daddr @private_ipv4_ips counter accept
-    ip6 saddr @allowed_ipv6_ips ip6 daddr #{guest_ephemeral} counter accept
-    ip saddr . tcp dport @allowed_ipv4_port_tuple ip daddr @private_ipv4_ips counter accept
+    ip saddr @allowed_ipv4_cidrs ip daddr @private_ipv4_cidrs counter accept
+    ip6 saddr @allowed_ipv6_cidrs ip6 daddr #{guest_ephemeral} counter accept
+    ip saddr . tcp dport @allowed_ipv4_port_tuple ip daddr @private_ipv4_cidrs counter accept
     ip6 saddr . tcp dport @allowed_ipv6_port_tuple ip6 daddr #{guest_ephemeral} counter accept
-    ip saddr . udp dport @allowed_ipv4_port_tuple ip daddr @private_ipv4_ips counter accept
+    ip saddr . udp dport @allowed_ipv4_port_tuple ip daddr @private_ipv4_cidrs counter accept
     ip6 saddr . udp dport @allowed_ipv6_port_tuple ip6 daddr #{guest_ephemeral} counter accept
   }
 }
 TEMPLATE
+
     pop "firewall rule is added"
   end
 
-  def generate_ip_port_set(rules)
-    rules.map do |rule|
-      ip = rule.ip.to_s
-      port_range = rule.port_range
-      port_rule = if port_range.begin == port_range.end - 1
-        port_range.begin.to_s
+  def consolidate_rules(rules)
```

Comment:
can you help me understand better the general undertaking this method has? Maybe it can be a comment, maybe I am just being thick right now.

Also, I'm wondering if the `#fill` method helps at all. It took me a long time to find and appreciate the fundamental nature of this method, but this code looks adjacent to its uses at least. https://rubydoc.info/gems/netaddr/NetAddr/IPv6Net#fill-instance_method

Reply:
I have typed up a comment. As a summary, the main problem here is to consolidate cidr_ranges + port_ranges. NetAddr methods like `fill` or `summ_IPv4/6Net` do not have port range support. I am still making use of the `summ` method but only after making sure those cidrs need to be summed up in that specific port range.

Reply:
btw, in general....am I just weird or do I expect vertical whitespace between paragraphs?

good comment though, I understand it way better now.

Reply:
I have updated it :)